### PR TITLE
Replace t5 Radius Matching with DNN + Pairwise Hit Differences + Make DNN First t5 Cut

### PR DESCRIPTION
**Pull request by GNiendorf**
Slides with PR Overview:  [PR #103 Overview.pdf](https://github.com/user-attachments/files/17382152/PR.103.Overview.pdf)

This PR Final Timing
Screenshot 2024-10-14 at 2 09 04 PM

Current Master Timing
372299297-a8a62749-1643-4540-b667-e23c7fc28d42

Replaces the t5 radius matching cuts with the DNN. As mentioned below, I kept the signal efficiency constant at 97.5% in each eta bin, so because I removed a cut with non-unity signal eff it makes sense that the efficiency increases slightly along with the FR. Marking this PR as ready for review.

Edit:

Pairwise hit differences help to decorrelate the input variables, speeding up training time by a factor of 4-5. Final AUC is largely the same, but faster training and less correlation among input features is useful.

Screenshot 2024-10-11 at 10 27 23 AM

Edit 2:

Using pragma unrolls results in a significant speedup of the DNN, resulting in an overall timing improvement from placing the DNN as the first t5 cut.

No unrolls, DNN before all cuts
Screenshot 2024-10-12 at 12 15 30 PM

With unrolls, DNN before all cuts
Screenshot 2024-10-12 at 12 17 35 PM

This PR Timing, prior to the unrolls optimization + moving DNN to front
Screenshot 2024-10-12 at 12 06 20 PM

Current Master Timing (taken from above):
372299297-a8a62749-1643-4540-b667-e23c7fc28d42

### CODE REVIEW

**Comment by GNiendorf** [comment]
I kept the signal efficiency constant, so the efficiency increases slightly along with the FR. Marking this PR as ready for review.

**Comment by GNiendorf** [comment]
Pairwise hit differences help to decorrelate the input variables, speeding up training time by a factor of 4-5. Final AUC is largely the same, but faster training and less correlation among input features is useful.

Screenshot 2024-10-11 at 10 27 23 AM


/run all

**Comment by slava77** [comment]
> I kept the signal efficiency constant, so the efficiency increases slightly along with the FR. Marking this PR as ready for review.

"constant" in which direction, vs eta or vs baseline?

**Comment by GNiendorf** [comment]
> > I kept the signal efficiency constant, so the efficiency increases slightly along with the FR. Marking this PR as ready for review.
> 
> "constant" in which direction, vs eta or vs baseline?

Sorry, I mean constant in eta (so I kept it at 97.5% signal efficiency in each bin). So because I removed a cut which had some non-unity signal efficiency, it makes sense that the overall t5/pt5 efficiency increase slightly along with the FR.

**Comment by GNiendorf** [comment]
Thank you for the comments Slava, working on them now. One additional optimization I just noticed is to use the eta value of one of the hits instead of the "t5_eta" variable since it was removed from the DNN, but I'll leave that to the next PR.

**Comment by GNiendorf** [comment]
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#pragma-unroll

I was playing around with some optimizations, and it seems like using pragma unroll in the linear layer and relu functions gives a small benefit? I'll leave it out of this PR for simplicity, but it might become useful later on.

Screenshot 2024-10-12 at 12 01 09 PM

**Comment by GNiendorf** [comment]
One interesting thing: If I put the DNN inference before the rz chi-squared and runQuintupletAlgoSelector cuts, I can actually get significantly faster performance with the unrolls @slava77 :

No unrolls
Screenshot 2024-10-12 at 12 15 30 PM

With unrolls
Screenshot 2024-10-12 at 12 17 35 PM

**Comment by GNiendorf** [comment]
I see a similar improvement on the L4 (above is L40)

Current Timing
Screenshot 2024-10-12 at 9 08 52 PM

With unrolls, DNN before all cuts
Screenshot 2024-10-12 at 9 06 23 PM

**Comment by GNiendorf** [comment]
Given the above, I think it makes sense to add the pragma's to this PR. Let me know if you think otherwise though @slava77 

/run all

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h
Comment (slava77) [inline RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:36]: are both similarly useful? I'd think the innermost is more significant

Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:36]: It seems the benefit is coming from the outermost pragma mainly.

Only innermost pragma
Screenshot 2024-10-14 at 8 17 31 AM

Only outermost pragma
Screenshot 2024-10-14 at 8 22 06 AM

File: RecoTracker/LSTCore/standalone/analysis/DNN/train_T5_DNN.ipynb
Comment (GNiendorf) [reply RecoTracker/LSTCore/standalone/analysis/DNN/train_T5_DNN.ipynb]: Fixed, thanks.

**Comment by GNiendorf** [comment]
Changed the cut bins to use the eta value of the first hit instead of t5_eta.

/run all

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h
Comment (slava77) [inline RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:41]:
```suggestion
  ALPAKA_FN_ACC ALPAKA_FN_INLINE bool runInference(TAcc const& acc,
```
to match what's actually returned

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h
Comment (slava77) [inline RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:83]: here and below, are we losing performance around eta or z around zero by taking deltas of abs values? Perhaps deltas of signed values are more useful?

File: RecoTracker/LSTCore/src/alpaka/Quintuplet.h
Comment (slava77) [inline RecoTracker/LSTCore/src/alpaka/Quintuplet.h:1995]: is it 5->`Params_T5::kLayers` here? or is it some different 5?
perhaps xVec, yVec can be explicitly sized as well

**Comment by GNiendorf** [comment]
It seems like only the final cut in these functions is contributing anything, I don't see any differences in performance from removing all but the dbeta cuts.

Timing Before Removing
Screenshot 2024-10-19 at 3 23 00 PM

Timing After Removing
Screenshot 2024-10-19 at 3 19 33 PM

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h
Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:80]: I just gave this a try but didn't see any improvement. Marking as resolved.

**Comment by GNiendorf** [comment]
@slava77 Given that only the dBeta and rz chi-squared cuts are left, i think it makes sense to leave this PR open to see if I can integrate those into the DNN before merging.

**Comment by slava77** [comment]
>  Run Standalone PU200 — Status: failure 

looks like due to rate limit https://www.docker.com/increase-rate-limits/

@ariostas 
is it possible that the 200/6 hours limit also has a lower period limit, like 1/minute to make sense out of these failures?
Perhaps some 30 sec sleep is needed in the CI

**Comment by ariostas** [comment]
> is it possible that the 200/6 hours limit also has a lower period limit, like 1/minute to make sense out of these failures?
> Perhaps some 30 sec sleep is needed in the CI

I'm not sure. I've always interpreted that as a client-side limitation. I'm guessing that that particular machine that was running the job happened to pull a bunch of Docker images recently

**Comment by slava77** [comment]
> There was a problem while building and running in standalone mode. The logs can be found [here](https://github.com/SegmentLinking/cmssw/actions/runs/11523134527).

```
make -C code/rooutil/
...
cxxopts.h:241:46: error: no match for 'operator+' (operand types are 'const char8_t [10]' and 'const std::string' {aka 'const std::__cxx11::basic_string<char>'})
  241 |         : OptionParseException(u8"Argument " + LQUOTE + text + RQUOTE + u8" starts with a - but has incorrect syntax") {}
```

is there a special symbol somewhere in RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc ?

**Comment by slava77** [comment]
> > There was a problem while building and running in standalone mode. The logs can be found [here](https://github.com/SegmentLinking/cmssw/actions/runs/11523134527).
> 
> ```
> make -C code/rooutil/
> ...
> cxxopts.h:241:46: error: no match for 'operator+' (operand types are 'const char8_t [10]' and 'const std::string' {aka 'const std::__cxx11::basic_string<char>'})
>   241 |         : OptionParseException(u8"Argument " + LQUOTE + text + RQUOTE + u8" starts with a - but has incorrect syntax") {}
> ```
> 
> is there a special symbol somewhere in RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc ?

..., this part is not compiling write_lst_ntuple; seems like some kind of glitch or is the CI variant having an issue

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h
Comment (slava77) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:14]: @GNiendorf 
do you see these warnings of ignored pragma in your tests as well?

**Comment by GNiendorf** [comment]
> > is there a special symbol somewhere in RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc ?
> > ..., this part is not compiling write_lst_ntuple; seems like some kind of glitch or is the CI variant having an issue
> 
> Oh this issue came up when we moved to C++20. I thought I had already fixed this in the devel branch

I haven't rebased my branch in the two weeks this PR has been open. Maybe that will fix things? Did you move the devel branch to C++ in that timeframe?

File: RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h
Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:14]: Yeah it looks like the CPU backend ignores the pragmas, I just tested it and they don't give a speedup on CPU.

Comment (GNiendorf) [reply RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:14]: Maybe I should wrap #pragma unroll in an ifdef only for the CUDA backend? I was told by Dan that CMSSW might have their own version of this as well?

**Comment by ariostas** [comment]
> I haven't rebased my branch in the two weeks this PR has been open. Maybe that will fix things? Did you move the devel branch to C++20 in that timeframe?

The PR merges before testing, so it must be that the devel branch hasn't been fixed.

**Comment by ariostas** [comment]
Hmm it's pretty weird because in https://github.com/SegmentLinking/cmssw/commit/1e6527f4ff98ed5d83469c7a378420bd47ee700f I did fix it for the devel branch. I wonder if it's not doing the right merge.

**Comment by ariostas** [comment]
I'm very confused, I'll just try again to see if it works.

/run standalone

**Comment by ariostas** [comment]
It doesn't make any sense. Here's part of the error:

```
cxxopts.h: In constructor 'cxxopts::Option_syntax_exception::Option_syntax_exception(const std::string&)':
cxxopts.h:241:46: error: no match for 'operator+' (operand types are 'const char8_t [10]' and 'const std::string' {aka 'const std::__cxx11::basic_string<char>'})
  241 |         : OptionParseException(u8"Argument " + LQUOTE + text + RQUOTE + u8" starts with a - but has incorrect syntax") {}
      |                                ~~~~~~~~~~~~~ ^ ~~~~~~
      |                                |               |
      |                                |               const std::string {aka const std::__cxx11::basic_string<char>}
      |                                const char8_t [10]
```

And here's the actual file in this branch.

https://github.com/SegmentLinking/cmssw/blob/782cf2fadc77b31b757c46124f4bc1a69451c52a/RecoTracker/LSTCore/standalone/code/rooutil/cxxopts.h#L238-L243

The `u8` is not there. However, it's showing that `ignoring '#pragma unroll FEATURES'` warning, so it's definitely using the right branch. I wonder if it's using some cached file or something.

**Comment by slava77** [comment]
> I wonder if it's using some cached file or something.

> And here's the actual file in this branch.
> 
> [cmssw/RecoTracker/LSTCore/standalone/code/rooutil/cxxopts.h](https://github.com/SegmentLinking/cmssw/blob/782cf2fadc77b31b757c46124f4bc1a69451c52a/RecoTracker/LSTCore/standalone/code/rooutil/cxxopts.h#L238-L243)

ah, but that's not the actual file, not this PR 782cf2fadc77b3, it should be the target branch, since this PR doesn't touch it: CMSSW_14_1_0_pre3_LST_X_LSTCore_realfiles_batch1_devel ends on 7a2c4228f537542fb2ea9497fb214b732c0bc245
https://github.com/SegmentLinking/cmssw/blob/7a2c4228f537542fb2ea9497fb214b732c0bc245/RecoTracker/LSTCore/standalone/code/rooutil/cxxopts.h#L239-L242

the change came from 992947d8d5a766089dd9e088c184bf71e95958ca

**Review by slava77 — APPROVED** [review]

**Pull request merged**
